src/test/demultiplexer_node_test.cpp

```diff
@@ -186,3 +186,105 @@ TEST(DemultiplexerTest, ShardsShareDataWithSourceTensor) {
         }
     }
 }
+
+// In case of demultiplexer right before gather, the input tensor object (TensorWithShape) will contain source tensor.
+// This test ensures the gathering prodcues tensor with correct data (which implies it takes care of source tensor ownership).
+TEST(DemultiplexerTest, GatherShardsWithExistingSourceTensors) {
+    // Mock DLNode session to be possible to get inputs from its input handler.
+    class MockNodeSession : public DLNodeSession {
+    public:
+        using DLNodeSession::DLNodeSession;
+
+        const TensorMap& getInputs() {
+            return this->inputHandler->getInputs();
+        }
+    };
+
+    // Mock DLNode to inject factory method to be able to create Node with mocked NodeSession.
+    class MockDLNode : public DLNode {
+    protected:
+        std::unique_ptr<NodeSession> createNodeSession(const NodeSessionMetadata& metadata, const CollapseDetails& collapsingDetails) override {
+            return std::make_unique<MockNodeSession>(metadata, getName(), previous.size(), collapsingDetails,
+                this->modelManager, this->modelName, this->modelVersion.value_or(0));
+        }
+
+    public:
+        using DLNode::DLNode;
+
+        const TensorMap& getInputs(const session_key_t& key) {
+            auto& nodeSession = dynamic_cast<MockNodeSession&>(this->getNodeSession(key));
+            return nodeSession.getInputs();
+        }
+    };
+
+    ConstructorEnabledModelManager manager;
+    int32_t demultiplyCount = 3;
+
+    auto dl_demulti = std::make_unique<DLNode>(
+        std::string("dummy_node_1"),
+        std::string("dummy_model"),
+        1, manager,
+        std::unordered_map<std::string, std::string>{{"b", "b"}},
+        demultiplyCount);
+    auto dl_gather = std::make_unique<MockDLNode>(
+        std::string("dummy_node_2"),
+        std::string("dummy_model"),
+        1, manager,
+        std::unordered_map<std::string, std::string>{{"b", "b"}},
+        std::nullopt,
+        std::set<std::string>{"dummy_node_1"});
+
+    dl_demulti->addDependant(*dl_gather);
+    dl_gather->addDependency(*dl_demulti, Aliases{{"b", "b"}});
+
+    size_t size = 2;
+
+    // Prepare source tensors
+    std::vector<float> data{3.2, 5.9};
+    std::vector<std::shared_ptr<ov::Tensor>> tensors;
+    for (int i = 0; i < demultiplyCount; i++) {
+        tensors.push_back(std::make_shared<ov::Tensor>(ov::element::f32, ov::Shape{1, size}));
+        std::memcpy(tensors[i]->data(), data.data(), sizeof(float) * size);
+    }
+
+    // Prepare session results and pass to ::setInputs method - source and actual tensors.
```

Comment:
Add comment that in this section we imitate demultiplication

Reply:
Done